pkg/kubelet/kubelet.go

```diff
@@ -1637,6 +1648,59 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 	return nil
 }
 
+func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, string) {
+	var otherActivePods []*v1.Pod
+	activePods := kl.GetActivePods()
+	for _, p := range activePods {
+		if p.UID != pod.UID {
+			otherActivePods = append(otherActivePods, p)
+		}
+	}
+	if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, pod); !ok {
+		// Log reason and return. Let the next sync iteration retry the resize
+		klog.V(2).Infof("Pod '%s' resize cannot be accommodated. Reason: '%s' Message: '%s'", pod.Name, failReason, failMessage)
+		return false, ""
+	}
+
+	var containersPatchData string
+	for _, container := range pod.Spec.Containers {
+		var resourcesPatchData string
+		for rName, rQuantity := range container.Resources.Requests {
+			container.ResourcesAllocated[rName] = rQuantity
+			resourcesPatchData += fmt.Sprintf(`"%s":"%s",`, rName, rQuantity.String())
+		}
+		resourcesPatchData = strings.TrimRight(resourcesPatchData, ",")
+		containersPatchData += fmt.Sprintf(`{"name":"%s","resourcesAllocated":{%s}},`, container.Name, resourcesPatchData)
+	}
+	kl.podManager.UpdatePod(pod)
```

Comment:
That's true. As of now, we are dealing with pointer to pod object that is being updated.
To clarify, are you suggesting we deepcopy the pod only for patch generation via strategicpatch, and continue to update 'pod' argument, and remove the UpdatePod? The pod object is consumed further down in SyncPod for updateContainerResources. It believe it will see another update from the apiserver patch which is inconsequential for CRI update, I'll verify it.

Reply:
I'd primarily like to be as explicit as possible in any updates we perform to the pod object.  I've had to debug mysterious "why did my pod object randomly change" kubelet bugs.  Given the size of the code base, and the large numbers of readers, it can be hard to figure out what could have possibly updated the pod object in the pod manager.

TBH, i'd prefer doing the deepcopy in the pod manager itself so that updates by readers are impossible, but that will likely cause a regression.  I don't want to couple that with your feature.  In lieu of that, lets just make the update as explicit as possible here.:

1. read the pod
2. deepcopy pod to newPod
3. modify newPod
4. patch(diff(pod, newPod)
5. podManager.UpdatePod(newPod)
6. pod = newPod
... continue on like before.

Reply:
yes, this is a reasonable compromise. i'll do it this way.

Reply:
fixed.